torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,637 @@
+
+  // Propagate transformations so far to the entire DAG
+  TransformPropagator propagator(reference1);
+  MaxRootDomainInfoSpanningTree entire_dag(reference1);
+  entire_dag.traverse(&propagator);
+  scheduler_utils::parallelizeAllLike(reference1);
+
+  // For a transpose scheduling, all we need is to bind threadIdx.x differently
+  // for inputs and outputs. This swap of binding could happen at any tensor on
+  // the path from input to output, especially, it does not have to be in the
+  // transpose tensor. Here, we naively do the binding swap at cached
+  // input/output for simplicity. We might need to find a better set of swap
+  // tensors in the future to reduce shared memory usage.
+
+  // transform tile for vectorization/unroll
+  // See note [vectorization and unroll of input and output]
+  int64_t num_elems_per_tile = params.tile_size1 * params.tile_size2;
+  int64_t num_elems_per_thread = num_elems_per_tile / kThreadsPerBlock;
+  int64_t unroll_factor1 = num_elems_per_thread / params.vectorize_factor1;
+  int64_t unroll_factor2 = num_elems_per_thread / params.vectorize_factor2;
+
+  int warp_size = at::cuda::getCurrentDeviceProperties()->warpSize;
+
+  // schedule group 2
+  int pos = reference2->nDims() - 2;
+  // [..., tile1, tile2]
+  reference2->merge(pos);
+  reference2->split(pos, params.vectorize_factor2);
+  reference2->split(pos, warp_size);
+  reference2->split(pos, unroll_factor2, false);
+  reference2->axis(-1)->parallelize(ParallelType::Vectorize);
+  reference2->axis(-2)->parallelize(ParallelType::TIDx);
+  reference2->axis(-3)->parallelize(ParallelType::TIDy);
+  reference2->axis(-4)->parallelize(ParallelType::Unroll);
+  // [..., Unroll, TIDy, TIDx, Vectorize]
+
+  // Propagate transformations of reference2 to the entire DAG except
+  // group 1. We actually only want to propagate to the fusion outputs, but
+  // inputs and outputs themselves are disconnected, so we have to borrow the
+  // entire DAG and use its spanning tree.
+  {
+    auto all_tvs_except1 = ir_utils::allTvsExcept(
+        fusion,
+        {grouped_inputs_outputs[0].begin(), grouped_inputs_outputs[0].end()});
+    SetSelector selector({all_tvs_except1.begin(), all_tvs_except1.end()});
+    MaxRootDomainInfoSpanningTree entire_dag_except1(reference2, &selector);
+    TransformPropagator propagator(reference2);
+    entire_dag_except1.traverse(&propagator);
+  }
+
+  // schedule group 1
+  reference1->reorder({{-2, -1}});
+  // [..., tile2, tile1]
+  pos = reference1->nDims() - 2;
+  reference1->merge(pos);
+  reference1->split(pos, params.vectorize_factor1);
+  reference1->split(pos, warp_size);
+  reference1->split(pos, unroll_factor1, false);
+  reference1->axis(-1)->parallelize(ParallelType::Vectorize);
+  reference1->axis(-2)->parallelize(ParallelType::TIDx);
+  reference1->axis(-3)->parallelize(ParallelType::TIDy);
+  reference1->axis(-4)->parallelize(ParallelType::Unroll);
+  // [..., Unroll, TIDy, TIDx, Vectorize]
```

Comment:
Seems like this approach would support non-square smem tiles of more or less any size.